File: 1001-1100/1040-Moving Stones Until Consecutive II/1040-Moving Stones Until Consecutive II.py
class Solution:
    def numMovesStonesII(self, stones: List[int]) -> List[int]:
        stones.sort()
        maxMove = max(stones[-2] - stones[0] - len(stones) + 2, stones[-1] - stones[1] - len(stones) + 2)
        minMove = len(stones)
        start = 0
        for i in range(len(stones)):
            while stones[i] - stones[start] >= len(stones):
                start += 1

            if i - start + 1 == len(stones) - 1 and stones[i] - stones[start] == len(stones) - 2:
                minMove = min(minMove, 2)
            else:
                minMove = min(minMove, len(stones) - (i - start + 1))

        return [minMove, maxMove]

File: 1001-1100/1040-Moving Stones Until Consecutive II/1040-Moving Stones Until Consecutive II.cpp
class Solution {
public:
    vector<int> numMovesStonesII(vector<int>& stones) {
        sort(stones.begin(), stones.end());
        int maxMove = max(stones[stones.size() - 2] + 2 - stones[0] - stones.size(),
                          stones.back() + 2 - stones[1] - stones.size());

        int minMove = stones.size();
        int start = 0;
        for (int i = 0; i < stones.size(); ++i) {
            while (stones[i] - stones[start] >= stones.size()) {
                ++start;
            }

            if (i - start + 1 == stones.size() - 1 && stones[i] - stones[start] == stones.size() - 2) {
                minMove = min(minMove, 2);
            }
            else {
                minMove = min(minMove, int(stones.size()) - (i - start + 1));
            }
        }

        return vector<int>{minMove, maxMove};
    }
};
